Stream Buffer Size

charcodelimit opened the pull request:
Workaround for the Clozure-CL maximum array size limit (array-total-size-limit) of (expt 2 24) elements

Introduced Buffer-Size argument to allow a dynamic adaption of the http-stream buffer size.
The small default stream buffer leads to high GC load for large downloads.

stassats commented:
Hardcoding 16777215 is not a good idea.

hanshuebner commented:
I'd like to know whether the special implementation for CCL is really
warranted, i.e. that it is actually substantially contributing to improved
performance.

2016-10-31 17:45 GMT+01:00 Stas Boukarev <notifications@github.com>:

> Hardcoding 16777215 is not a good idea.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/edicl/drakma/pull/75#issuecomment-257348172>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGozzJRY--KTJuWGG9cMkn2wH68ljZSks5q5hsdgaJpZM4KlMSF>
> .
>

stassats commented:
And the docstring in %read-body is broken.

charcodelimit commented:
... the special implementation for CCL is no performance, but a 
correctness issue:

    > Error: Cannot allocate a (CCL::UVECTOR 199) with 18734900 elements.
    >        Objects of type (CCL::UVECTOR 199) can can have at most
    >        16777215 elements in this implementation.


On 31.10.2016 17:46, Hans Hübner wrote:
> I'd like to know whether the special implementation for CCL is really
> warranted, i.e. that it is actually substantially contributing to improved
> performance.
>
> 2016-10-31 17:45 GMT+01:00 Stas Boukarev <notifications@github.com>:
>
> > Hardcoding 16777215 is not a good idea.
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/edicl/drakma/pull/75#issuecomment-257348172>, or 
> mute
> > the thread
> > 
> <https://github.com/notifications/unsubscribe-auth/AAGozzJRY--KTJuWGG9cMkn2wH68ljZSks5q5hsdgaJpZM4KlMSF>
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/edicl/drakma/pull/75#issuecomment-257348667>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/APn5KFNhZ_K5oLBmY8SNZfiVehlmFN4Qks5q5huCgaJpZM4KlMSF>.
>

hanshuebner commented:
I don't understand why a different loop is used for ccl than for the other
implementations.  If this is a requirement, then there should be a comment
which clearly describes why the code that is used for all other
implementations cannot be used for ccl 32bit.

2016-10-31 17:56 GMT+01:00 C. Hofmann-Fuchs <notifications@github.com>:

> ... the special implementation for CCL is no performance, but a
> correctness issue:
>
> > Error: Cannot allocate a (CCL::UVECTOR 199) with 18734900 elements.
> > Objects of type (CCL::UVECTOR 199) can can have at most
> > 16777215 elements in this implementation.
>
>
> On 31.10.2016 17:46, Hans Hübner wrote:
> > I'd like to know whether the special implementation for CCL is really
> > warranted, i.e. that it is actually substantially contributing to
> improved
> > performance.
> >
> > 2016-10-31 17:45 GMT+01:00 Stas Boukarev <notifications@github.com>:
> >
> > > Hardcoding 16777215 is not a good idea.
> > >
> > > —
> > > You are receiving this because you are subscribed to this thread.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/edicl/drakma/pull/75#issuecomment-257348172>, or
> > mute
> > > the thread
> > >
> > <https://github.com/notifications/unsubscribe-auth/AAGozzJRY--
> KTJuWGG9cMkn2wH68ljZSks5q5hsdgaJpZM4KlMSF>
> > > .
> > >
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/edicl/drakma/pull/75#issuecomment-257348667>, or
> > mute the thread
> > <https://github.com/notifications/unsubscribe-auth/APn5KFNhZ_
> K5oLBmY8SNZfiVehlmFN4Qks5q5huCgaJpZM4KlMSF>.
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/edicl/drakma/pull/75#issuecomment-257351723>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGoz2XdnY6z3WRKJpp8n5CsvXQ-LoNtks5q5h3VgaJpZM4KlMSF>
> .
>

stassats commented:
It returns a list for the CCL version, does that actually work?

stassats commented:
And it doesn't.

stassats commented:
And for the other part of the PR, how much does changing the buffer size actually help? Can we instead increase the default value? I'd prefer not to needlessly complicate things.

charcodelimit commented:
... here are some benchmark results, although it took me a while.

Plain-flavor HTTP-streams are no problem, but the difference in speed for
chunked HTTP-streams is roughly factor 30.

Attached you will find what SBCL's profiler has to say about it.


(let ((body) (status) (headers)
       (uri) (http-stream) (must-close)
       (status-text))
   (time
    (multiple-value-setq (body status headers uri http-stream must-close 
status-text)
        (drakma:http-request (puri:parse-uri 
"http://www.test.internal/cgi-bin/get-file.pl")
                             :preserve-uri t
                             :external-format-out :UTF-8
                             :force-binary t)))
   (format t "~%Status: ~A" headers))

(MULTIPLE-VALUE-SETQ (BODY STATUS HEADERS URI HTTP-STREAM MUST-CLOSE 
STATUS-TEXT) (DRAKMA:HTTP-REQUEST (PURI:PARSE-URI 
"http://www.test.internal/cgi-bin/get-file.pl") :PRESERVE-URI T 
:EXTERNAL-FORMAT-OUT :UTF-8 :FORCE-BINARY T))
took 631,233,402 microseconds (631.233400 seconds) to run.
       62,013,753 microseconds ( 62.013752 seconds, 9.82%) of which was 
spent in GC.
During that period, and with 4 available CPU cores,
      633,284,000 microseconds (633.284000 seconds) were spent in user mode
                0 microseconds (  0.000000 seconds) were spent in system 
mode
  57,309,094,128 bytes of memory allocated.
  13,658 minor page faults, 0 major page faults, 0 swaps.

Status: ((DATE . Tue, 01 Nov 2016 08:14:58 GMT) (SERVER . Apache) 
(X-FRAME-OPTIONS . SAMEORIGIN,SAMEORIGIN) (X-CONTENT-TYPE-OPTIONS . 
nosniff,nosniff) (X-XSS-PROTECTION . 1; mode=block,1; mode=block) 
(CONTENT-TRANSFER-ENCODING . binary) (CONTENT-DISPOSITION . attachment; 
filename="test.dat") (CONTENT-DESCRIPTION . dat file) (CONTENT-TYPE . 
application/octet-stream) (VIA . 1.1 www.test.internal-80) 
(CACHE-CONTROL . max-age=14400) (EXPIRES . Tue, 01 Nov 2016 12:14:58 
GMT) (CONNECTION . close) (TRANSFER-ENCODING . chunked))
NIL

(let ((body) (status) (headers)
       (uri) (http-stream) (must-close)
       (status-text))
   (time
    (multiple-value-setq (body status headers uri http-stream must-close 
status-text)
        (drakma:http-request (puri:parse-uri 
"http://www.test.internal/cgi-bin/get-file.pl")
                             :preserve-uri t
                             :external-format-out :UTF-8
                             :force-binary t
                             :buffer-size (ash 1 20))))
   (format t "~%Status: ~A" headers))

NIL

(MULTIPLE-VALUE-SETQ (BODY STATUS HEADERS URI HTTP-STREAM MUST-CLOSE 
STATUS-TEXT) (DRAKMA:HTTP-REQUEST (PURI:PARSE-URI 
"http://www.test.internal/cgi-bin/get-file.pl") :PRESERVE-URI T 
:EXTERNAL-FORMAT-OUT :UTF-8 :FORCE-BINARY T :BUFFER-SIZE (ASH 1 20)))
took 19,894,252 microseconds (19.894253 seconds) to run.
       1,035,845 microseconds ( 1.035845 seconds, 5.21%) of which was 
spent in GC.
During that period, and with 4 available CPU cores,
       8,920,000 microseconds ( 8.920000 seconds) were spent in user mode
               0 microseconds ( 0.000000 seconds) were spent in system mode
  488,198,528 bytes of memory allocated.
  7,088 minor page faults, 0 major page faults, 0 swaps.

Status: ((DATE . Tue, 01 Nov 2016 08:26:15 GMT) (SERVER . Apache) 
(X-FRAME-OPTIONS . SAMEORIGIN,SAMEORIGIN) (X-CONTENT-TYPE-OPTIONS . 
nosniff,nosniff) (X-XSS-PROTECTION . 1; mode=block,1; mode=block) 
(CONTENT-TRANSFER-ENCODING . binary) (CONTENT-DISPOSITION . attachment; 
filename="test.dat") (CONTENT-DESCRIPTION . dat file) (CONTENT-TYPE . 
application/octet-stream) (VIA . 1.1 www.test.internal-80) 
(CACHE-CONTROL . max-age=14400) (EXPIRES . Tue, 01 Nov 2016 12:26:15 
GMT) (CONNECTION . close) (TRANSFER-ENCODING . chunked))
NIL

;;;;;;;;;;;;;;;;;;;;;;

It should be possible to reproduce the poor performance with something 
along the following lines:

;; http://www.httpwatch.com/httpgallery/chunked/

(multiple-value-bind (body status headers uri http-stream must-close 
status-text)
  (drakma:http-request (puri:parse-uri 
"http://www.httpwatch.com/httpgallery/chunked/chunkedimage.aspx")
                            :preserve-uri t
                            :external-format-out :UTF-8
                            :force-binary t
                            :buffer-size 2)))

Regards,
  Christian


On 31.10.2016 21:00, Stas Boukarev wrote:
>
> And for the other part of the PR, how much does changing the buffer 
> size actually help? Can we instead increase the default value? I'd 
> prefer not to needlessly complicate things.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/edicl/drakma/pull/75#issuecomment-257403607>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/APn5KHm-E_KjWoAWdSbEvdO5dcmYSTFbks5q5kjlgaJpZM4KlMSF>.
>

charcodelimit commented:
Depend's on what you want to do with it ...

The following works for me:

defun test-1 ()
   (let ((body) (status) (headers)
         (uri) (http-stream) (must-close)
         (status-text))
     (time
      (dotimes (i 10)
        (multiple-value-setq (body status headers uri http-stream 
must-close status-text)
            (drakma:http-request (puri:parse-uri 
"http://www.test.local/chinual_4thEd_Jul81.pdf")
                                 :preserve-uri t
                                 :external-format-out :UTF-8
                                 :force-binary t
                                 :buffer-size (ash 1 26)))))
     (format t "~%Status: ~A" headers)
     (with-open-file (out "chineual-local.pdf"
                          :element-type '(unsigned-byte 8)
                          :direction :output
                          :if-does-not-exist :create
                          :if-exists :supersede)
       (write-sequence body out))
     nil))

Regards,
  Christian

On 31.10.2016 20:59, Stas Boukarev wrote:
>
> And it doesn't.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/edicl/drakma/pull/75#issuecomment-257403206>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/APn5KJrK4ULE6DTK2juvUlspZ_xW6Ntrks5q5kiMgaJpZM4KlMSF>.
>

hanshuebner commented:
Please provide a synopsis of your findings.  Also describe why the approach
that works well on ccl 32bit does not work for other implementations.  I am
unwilling to merge a change that adds a different algorithm for a basically
obsolete architecture. Performance improvements are welcome, but we do need
to consider maintainability as well.

2016-11-01 20:45 GMT+01:00 C. Hofmann-Fuchs <notifications@github.com>:

> Depend's on what you want to do with it ...
>
> The following works for me:
>
> defun test-1 ()
> (let ((body) (status) (headers)
> (uri) (http-stream) (must-close)
> (status-text))
> (time
> (dotimes (i 10)
> (multiple-value-setq (body status headers uri http-stream
> must-close status-text)
> (drakma:http-request (puri:parse-uri
> "http://www.test.local/chinual_4thEd_Jul81.pdf")
> :preserve-uri t
> :external-format-out :UTF-8
> :force-binary t
> :buffer-size (ash 1 26)))))
> (format t "~%Status: ~A" headers)
> (with-open-file (out "chineual-local.pdf"
> :element-type '(unsigned-byte 8)
> :direction :output
> :if-does-not-exist :create
> :if-exists :supersede)
> (write-sequence body out))
> nil))
>
> Regards,
> Christian
>
> On 31.10.2016 20:59, Stas Boukarev wrote:
> >
> > And it doesn't.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/edicl/drakma/pull/75#issuecomment-257403206>, or
> > mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> APn5KJrK4ULE6DTK2juvUlspZ_xW6Ntrks5q5kiMgaJpZM4KlMSF>.
>
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/edicl/drakma/pull/75#issuecomment-257673331>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGoz46joSDNR4nFv3p7BI37yCigJn86ks5q55bAgaJpZM4KlMSF>
> .
>

charcodelimit commented:
Synopsis:
  - reading chunked HTTP-streams with a too small buffer-size leads to 
30x worse performance
    on all several CL-implementations (SBCL 1.3.9 X86-64, CCL 1.12 
X86-64, ECL 13.5.1 X86-64,  sbcl-1.3.9-armhf)

I can understand your argument about ccl 32-bit, even though I wouldn't call
the armv7 architecture obsolete which was a reason why I fixed this in 
the first place.

Regards,
  Christian

On 01.11.2016 20:48, Hans Hübner wrote:
> Please provide a synopsis of your findings. Also describe why the approach
> that works well on ccl 32bit does not work for other implementations. I am
> unwilling to merge a change that adds a different algorithm for a 
> basically
> obsolete architecture. Performance improvements are welcome, but we do 
> need
> to consider maintainability as well.
>
> 2016-11-01 20:45 GMT+01:00 C. Hofmann-Fuchs <notifications@github.com>:
>
> > Depend's on what you want to do with it ...
> >
> > The following works for me:
> >
> > defun test-1 ()
> > (let ((body) (status) (headers)
> > (uri) (http-stream) (must-close)
> > (status-text))
> > (time
> > (dotimes (i 10)
> > (multiple-value-setq (body status headers uri http-stream
> > must-close status-text)
> > (drakma:http-request (puri:parse-uri
> > "http://www.test.local/chinual_4thEd_Jul81.pdf")
> > :preserve-uri t
> > :external-format-out :UTF-8
> > :force-binary t
> > :buffer-size (ash 1 26)))))
> > (format t "~%Status: ~A" headers)
> > (with-open-file (out "chineual-local.pdf"
> > :element-type '(unsigned-byte 8)
> > :direction :output
> > :if-does-not-exist :create
> > :if-exists :supersede)
> > (write-sequence body out))
> > nil))
> >
> > Regards,
> > Christian
> >
> > On 31.10.2016 20:59, Stas Boukarev wrote:
> > >
> > > And it doesn't.
> > >
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/edicl/drakma/pull/75#issuecomment-257403206>, or
> > > mute the thread
> > > <https://github.com/notifications/unsubscribe-auth/
> > APn5KJrK4ULE6DTK2juvUlspZ_xW6Ntrks5q5kiMgaJpZM4KlMSF>.
> >
> > >
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/edicl/drakma/pull/75#issuecomment-257673331>, or 
> mute
> > the thread
> > 
> <https://github.com/notifications/unsubscribe-auth/AAGoz46joSDNR4nFv3p7BI37yCigJn86ks5q55bAgaJpZM4KlMSF>
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/edicl/drakma/pull/75#issuecomment-257674428>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/APn5KI3PRHGejL7tPOSLHS5KnIMOtUuvks5q55eggaJpZM4KlMSF>.
>

hanshuebner commented:
Please use your better-performing loop as replacement for the existing
loop, for all platforms.

2016-11-01 21:06 GMT+01:00 C. Hofmann-Fuchs <notifications@github.com>:

> Synopsis:
> - reading chunked HTTP-streams with a too small buffer-size leads to
> 30x worse performance
> on all several CL-implementations (SBCL 1.3.9 X86-64, CCL 1.12
> X86-64, ECL 13.5.1 X86-64, sbcl-1.3.9-armhf)
>
> I can understand your argument about ccl 32-bit, even though I wouldn't
> call
> the armv7 architecture obsolete which was a reason why I fixed this in
> the first place.
>
> Regards,
> Christian
>
>
> On 01.11.2016 20:48, Hans Hübner wrote:
> > Please provide a synopsis of your findings. Also describe why the
> approach
> > that works well on ccl 32bit does not work for other implementations. I
> am
> > unwilling to merge a change that adds a different algorithm for a
> > basically
> > obsolete architecture. Performance improvements are welcome, but we do
> > need
> > to consider maintainability as well.
> >
> > 2016-11-01 20:45 GMT+01:00 C. Hofmann-Fuchs <notifications@github.com>:
> >
> > > Depend's on what you want to do with it ...
> > >
> > > The following works for me:
> > >
> > > defun test-1 ()
> > > (let ((body) (status) (headers)
> > > (uri) (http-stream) (must-close)
> > > (status-text))
> > > (time
> > > (dotimes (i 10)
> > > (multiple-value-setq (body status headers uri http-stream
> > > must-close status-text)
> > > (drakma:http-request (puri:parse-uri
> > > "http://www.test.local/chinual_4thEd_Jul81.pdf")
> > > :preserve-uri t
> > > :external-format-out :UTF-8
> > > :force-binary t
> > > :buffer-size (ash 1 26)))))
> > > (format t "~%Status: ~A" headers)
> > > (with-open-file (out "chineual-local.pdf"
> > > :element-type '(unsigned-byte 8)
> > > :direction :output
> > > :if-does-not-exist :create
> > > :if-exists :supersede)
> > > (write-sequence body out))
> > > nil))
> > >
> > > Regards,
> > > Christian
> > >
> > > On 31.10.2016 20:59, Stas Boukarev wrote:
> > > >
> > > > And it doesn't.
> > > >
> > > > —
> > > > You are receiving this because you authored the thread.
> > > > Reply to this email directly, view it on GitHub
> > > > <https://github.com/edicl/drakma/pull/75#issuecomment-257403206>, or
> > > > mute the thread
> > > > <https://github.com/notifications/unsubscribe-auth/
> > > APn5KJrK4ULE6DTK2juvUlspZ_xW6Ntrks5q5kiMgaJpZM4KlMSF>.
> > >
> > > >
> > >
> > > —
> > > You are receiving this because you commented.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/edicl/drakma/pull/75#issuecomment-257673331>, or
> > mute
> > > the thread
> > >
> > <https://github.com/notifications/unsubscribe-auth/
> AAGoz46joSDNR4nFv3p7BI37yCigJn86ks5q55bAgaJpZM4KlMSF>
> > > .
> > >
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/edicl/drakma/pull/75#issuecomment-257674428>, or
> > mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> APn5KI3PRHGejL7tPOSLHS5KnIMOtUuvks5q55eggaJpZM4KlMSF>.
>
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/edicl/drakma/pull/75#issuecomment-257679591>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGoz8Zr-I7UFakMroq7gimLGvptIRBTks5q55utgaJpZM4KlMSF>
> .
>

charcodelimit commented:
O.K. I changed the code accordingly

The tests look O.K. (see attachment for SBCL 1.3.9 X86-64, CCL 1.12 
X86-64, ECL 13.5.1 X86-64)


On 01.11.2016 21:08, Hans Hübner wrote:
> Please use your better-performing loop as replacement for the existing
> loop, for all platforms.
>
> 2016-11-01 21:06 GMT+01:00 C. Hofmann-Fuchs <notifications@github.com>:
>
> > Synopsis:
> > - reading chunked HTTP-streams with a too small buffer-size leads to
> > 30x worse performance
> > on all several CL-implementations (SBCL 1.3.9 X86-64, CCL 1.12
> > X86-64, ECL 13.5.1 X86-64, sbcl-1.3.9-armhf)
> >
> > I can understand your argument about ccl 32-bit, even though I wouldn't
> > call
> > the armv7 architecture obsolete which was a reason why I fixed this in
> > the first place.
> >
> > Regards,
> > Christian
> >
> >
> > On 01.11.2016 20:48, Hans Hübner wrote:
> > > Please provide a synopsis of your findings. Also describe why the
> > approach
> > > that works well on ccl 32bit does not work for other 
> implementations. I
> > am
> > > unwilling to merge a change that adds a different algorithm for a
> > > basically
> > > obsolete architecture. Performance improvements are welcome, but we do
> > > need
> > > to consider maintainability as well.
> > >
> > > 2016-11-01 20:45 GMT+01:00 C. Hofmann-Fuchs 
> <notifications@github.com>:
> > >
> > > > Depend's on what you want to do with it ...
> > > >
> > > > The following works for me:
> > > >
> > > > defun test-1 ()
> > > > (let ((body) (status) (headers)
> > > > (uri) (http-stream) (must-close)
> > > > (status-text))
> > > > (time
> > > > (dotimes (i 10)
> > > > (multiple-value-setq (body status headers uri http-stream
> > > > must-close status-text)
> > > > (drakma:http-request (puri:parse-uri
> > > > "http://www.test.local/chinual_4thEd_Jul81.pdf")
> > > > :preserve-uri t
> > > > :external-format-out :UTF-8
> > > > :force-binary t
> > > > :buffer-size (ash 1 26)))))
> > > > (format t "~%Status: ~A" headers)
> > > > (with-open-file (out "chineual-local.pdf"
> > > > :element-type '(unsigned-byte 8)
> > > > :direction :output
> > > > :if-does-not-exist :create
> > > > :if-exists :supersede)
> > > > (write-sequence body out))
> > > > nil))
> > > >
> > > > Regards,
> > > > Christian
> > > >
> > > > On 31.10.2016 20:59, Stas Boukarev wrote:
> > > > >
> > > > > And it doesn't.
> > > > >
> > > > > —
> > > > > You are receiving this because you authored the thread.
> > > > > Reply to this email directly, view it on GitHub
> > > > > 
> <https://github.com/edicl/drakma/pull/75#issuecomment-257403206>, or
> > > > > mute the thread
> > > > > <https://github.com/notifications/unsubscribe-auth/
> > > > APn5KJrK4ULE6DTK2juvUlspZ_xW6Ntrks5q5kiMgaJpZM4KlMSF>.
> > > >
> > > > >
> > > >
> > > > —
> > > > You are receiving this because you commented.
> > > > Reply to this email directly, view it on GitHub
> > > > <https://github.com/edicl/drakma/pull/75#issuecomment-257673331>, or
> > > mute
> > > > the thread
> > > >
> > > <https://github.com/notifications/unsubscribe-auth/
> > AAGoz46joSDNR4nFv3p7BI37yCigJn86ks5q55bAgaJpZM4KlMSF>
> > > > .
> > > >
> > >
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/edicl/drakma/pull/75#issuecomment-257674428>, or
> > > mute the thread
> > > <https://github.com/notifications/unsubscribe-auth/
> > APn5KI3PRHGejL7tPOSLHS5KnIMOtUuvks5q55eggaJpZM4KlMSF>.
> >
> > >
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/edicl/drakma/pull/75#issuecomment-257679591>, or 
> mute
> > the thread
> > 
> <https://github.com/notifications/unsubscribe-auth/AAGoz8Zr-I7UFakMroq7gimLGvptIRBTks5q55utgaJpZM4KlMSF>
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/edicl/drakma/pull/75#issuecomment-257680234>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/APn5KI3RuTBjKpChTec9J_LO1_o2eFtwks5q55w-gaJpZM4KlMSF>.
>

hanshuebner commented on request.lisp:
```diff
@@ -118,35 +118,41 @@ body using the boundary BOUNDARY."
+  ;; is (expt 2 24) elements.
+  ;; See: http://ccl.clozure.com/manual/chapter4.7.html
+  ;; Use a list as buffer
+  (progn
```
This `progn` is no longer needed.

hanshuebner commented on request.lisp:
```diff
@@ -118,35 +118,41 @@ body using the boundary BOUNDARY."
+    (when (>= buffer-size array-total-size-limit)
+      (setq buffer-size (- array-total-size-limit 1)))
+    (let ((buffer (make-array buffer-size :element-type element-type))
+          (result ()))
```
Both `buffer` and `result` can be declared inside of the loop using `with`.

hanshuebner commented on request.lisp:
```diff
@@ -118,35 +118,41 @@ body using the boundary BOUNDARY."
+      (setq buffer-size (- array-total-size-limit 1)))
+    (let ((buffer (make-array buffer-size :element-type element-type))
+          (result ()))
+      (loop :for index = 0 :then (+ index pos)
```
Please use bare symbols, not keywords, as loop keywords.

hanshuebner commented on request.lisp:
```diff
@@ -118,35 +118,41 @@ body using the boundary BOUNDARY."
+  ;; Use a list as buffer
+  (progn
+    (when (>= buffer-size array-total-size-limit)
+      (setq buffer-size (- array-total-size-limit 1)))
```
Re-bind buffer size instead of using `setq`.  Do so inside of the `loop` using `with` (see below for more of that).

hanshuebner commented on request.lisp:
```diff
@@ -118,35 +118,41 @@ body using the boundary BOUNDARY."
+          (result ()))
+      (loop :for index = 0 :then (+ index pos)
+            :for pos = (read-sequence buffer stream :start 0 :end buffer-size)
+            :do (dotimes (i pos) (push (aref buffer i) result))
```
Put the `push` onto a separate line.

stassats commented:
Still broken. You can't return a list from %read-body, it's asking for a string.

charcodelimit commented:
You mean something like this?

(loop with array-size (if (>= buffer-size array-total-size-limit)
                             (- array-total-size-limit 1)
                             buffer-size)
         with buffer = (make-array buffer-size :element-type element-type
         with result = ()
         for index = 0 then (+ index pos)
         for pos = (read-sequence buffer stream :start 0 :end buffer-size)
         do (dotimes (i pos)
              (push (aref buffer i) result))
         while (= pos array-size))
     (nreverse result))

Of cause you are right with removing the progn.
But above loop looks to me more intimidating than before.

One could even argue that the following version would be more consistent 
in the use of the loop macro:

(loop with array-size (if (>= buffer-size array-total-size-limit)
                             (- array-total-size-limit 1)
                             buffer-size)
       with buffer = (make-array buffer-size :element-type element-type
       with result = ()
       for index = 0 then (+ index pos)
       for pos = (read-sequence buffer stream :start 0 :end buffer-size)
       do (loop for i from 0 below pos
                collecting (aref buffer i) into result)
       while (= pos array-size)))

What is your opinion?

Regards,
  Christian

On 01.11.2016 22:20, Hans Hübner wrote:
>
> *@hanshuebner* commented on this pull request.
>
> Just a number of minor stylistic comments now. Looks good otherwise.
>
> ------------------------------------------------------------------------
>
> In request.lisp 
> <https://github.com/edicl/drakma/pull/75#pullrequestreview-6707176>:
>
> > -  "Helper function to read from stream into a buffer of element-type, which is returned."
> -  (let ((buffer (make-array +buffer-size+ :element-type element-type))
> -        (result (make-array 0 :element-type element-type :adjustable t)))
> -        (loop for index = 0 then (+ index pos)
> -           for pos = (read-sequence buffer stream)
> -           do (adjust-array result (+ index pos))
> -             (replace result buffer :start1 index :end2 pos)
> -           while (= pos +buffer-size+))
> -        result))
> -
> -(defun read-body (stream headers textp &key (decode-content t))
> +  ;; In 32-bit Clozure-CL the maximum array size limit
> +  ;; is (expt 2 24) elements.
> +  ;; See: http://ccl.clozure.com/manual/chapter4.7.html
> +  ;; Use a list as buffer
> +  (progn
>
> This |progn| is no longer needed.
>
> ------------------------------------------------------------------------
>
> In request.lisp 
> <https://github.com/edicl/drakma/pull/75#pullrequestreview-6707176>:
>
> > -           for pos = (read-sequence buffer stream)
> -           do (adjust-array result (+ index pos))
> -             (replace result buffer :start1 index :end2 pos)
> -           while (= pos +buffer-size+))
> -        result))
> -
> -(defun read-body (stream headers textp &key (decode-content t))
> +  ;; In 32-bit Clozure-CL the maximum array size limit
> +  ;; is (expt 2 24) elements.
> +  ;; See: http://ccl.clozure.com/manual/chapter4.7.html
> +  ;; Use a list as buffer
> +  (progn
> +    (when (>= buffer-size array-total-size-limit)
> +      (setq buffer-size (- array-total-size-limit 1)))
> +    (let ((buffer (make-array buffer-size :element-type element-type))
> +          (result ()))
>
> Both |buffer| and |result| can be declared inside of the loop using 
> |with|.
>
> ------------------------------------------------------------------------
>
> In request.lisp 
> <https://github.com/edicl/drakma/pull/75#pullrequestreview-6707176>:
>
> > -           do (adjust-array result (+ index pos))
> -             (replace result buffer :start1 index :end2 pos)
> -           while (= pos +buffer-size+))
> -        result))
> -
> -(defun read-body (stream headers textp &key (decode-content t))
> +  ;; In 32-bit Clozure-CL the maximum array size limit
> +  ;; is (expt 2 24) elements.
> +  ;; See: http://ccl.clozure.com/manual/chapter4.7.html
> +  ;; Use a list as buffer
> +  (progn
> +    (when (>= buffer-size array-total-size-limit)
> +      (setq buffer-size (- array-total-size-limit 1)))
> +    (let ((buffer (make-array buffer-size :element-type element-type))
> +          (result ()))
> +      (loop :for index = 0 :then (+ index pos)
>
> Please use bare symbols, not keywords, as loop keywords.
>
> ------------------------------------------------------------------------
>
> In request.lisp 
> <https://github.com/edicl/drakma/pull/75#pullrequestreview-6707176>:
>
> > -        (result (make-array 0 :element-type element-type :adjustable t)))
> -        (loop for index = 0 then (+ index pos)
> -           for pos = (read-sequence buffer stream)
> -           do (adjust-array result (+ index pos))
> -             (replace result buffer :start1 index :end2 pos)
> -           while (= pos +buffer-size+))
> -        result))
> -
> -(defun read-body (stream headers textp &key (decode-content t))
> +  ;; In 32-bit Clozure-CL the maximum array size limit
> +  ;; is (expt 2 24) elements.
> +  ;; See: http://ccl.clozure.com/manual/chapter4.7.html
> +  ;; Use a list as buffer
> +  (progn
> +    (when (>= buffer-size array-total-size-limit)
> +      (setq buffer-size (- array-total-size-limit 1)))
>
> Re-bind buffer size instead of using |setq|. Do so inside of the 
> |loop| using |with| (see below for more of that).
>
> ------------------------------------------------------------------------
>
> In request.lisp 
> <https://github.com/edicl/drakma/pull/75#pullrequestreview-6707176>:
>
> > -           while (= pos +buffer-size+))
> -        result))
> -
> -(defun read-body (stream headers textp &key (decode-content t))
> +  ;; In 32-bit Clozure-CL the maximum array size limit
> +  ;; is (expt 2 24) elements.
> +  ;; See: http://ccl.clozure.com/manual/chapter4.7.html
> +  ;; Use a list as buffer
> +  (progn
> +    (when (>= buffer-size array-total-size-limit)
> +      (setq buffer-size (- array-total-size-limit 1)))
> +    (let ((buffer (make-array buffer-size :element-type element-type))
> +          (result ()))
> +      (loop :for index = 0 :then (+ index pos)
> +            :for pos = (read-sequence buffer stream :start 0 :end buffer-size)
> +            :do (dotimes (i pos) (push (aref buffer i) result))
>
> Put the |push| onto a separate line.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/edicl/drakma/pull/75#pullrequestreview-6707176>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/APn5KN4Wo09h4wYtk-F3a7eB3wmU0gk2ks5q560xgaJpZM4KlMSF>.
>

charcodelimit commented:
Could you please add a unit test for this case.

On 01.11.2016 23:08, Stas Boukarev wrote:
>
> Still broken. You can't return a list from %read-body, it's asking for 
> a string.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/edicl/drakma/pull/75#issuecomment-257713977>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/APn5KE5Wfao4Q0z_XBKN1DSTLkPso00wks5q57htgaJpZM4KlMSF>.
>

charcodelimit commented:
I meant of cause the following alternative implementation using only the 
loop macro:

(defun %read-body (stream element-type &key (buffer-size +buffer-size+))
   "Helper function to read from stream into a buffer of element-type, 
which is returned."
   ;; On ABCL, a flexi-stream is not a normal stream. This is caused by
   ;; a bug in ABCL which is supposedly quite difficult to fix. More
   ;; details here: http://abcl.org/trac/ticket/377
   #-abcl
   (declare (stream stream))
   ;; In 32-bit Clozure-CL the maximum array size limit
   ;; is (expt 2 24) elements.
   ;; See: http://ccl.clozure.com/manual/chapter4.7.html
   ;; Use a list as buffer
   (let ((result ()))
     (setq result
           (loop with array-size = (if (>= buffer-size 
array-total-size-limit)
                                       (- array-total-size-limit 1)
                                       buffer-size)
                 with buffer = (make-array array-size :element-type 
element-type)
                 for index = 0 then (+ index pos)
                 for pos = (read-sequence buffer stream :start 0 :end 
array-size)
                 nconcing (loop for i from 0 below pos
                                collect (aref buffer i))
                 while (= pos array-size)))
     (if (subtypep element-type 'character)
         (coerce result 'string)
         result)))

On 02.11.2016 20:02, Christian Hofmann wrote:
> You mean something like this?
>
> (loop with array-size (if (>= buffer-size array-total-size-limit)
>                             (- array-total-size-limit 1)
>                             buffer-size)
>         with buffer = (make-array buffer-size :element-type element-type
>         with result = ()
>         for index = 0 then (+ index pos)
>         for pos = (read-sequence buffer stream :start 0 :end buffer-size)
>         do (dotimes (i pos)
>              (push (aref buffer i) result))
>         while (= pos array-size))
>     (nreverse result))
>
> Of cause you are right with removing the progn.
> But above loop looks to me more intimidating than before.
>
> One could even argue that the following version would be more 
> consistent in the use of the loop macro:
>
> (loop with array-size (if (>= buffer-size array-total-size-limit)
>                             (- array-total-size-limit 1)
>                             buffer-size)
>       with buffer = (make-array buffer-size :element-type element-type
>       with result = ()
>       for index = 0 then (+ index pos)
>       for pos = (read-sequence buffer stream :start 0 :end buffer-size)
>       do (loop for i from 0 below pos
>                collecting (aref buffer i) into result)
>       while (= pos array-size)))
>
> What is your opinion?
>
> Regards,
>  Christian
>
> On 01.11.2016 22:20, Hans Hübner wrote:
>>
>> *@hanshuebner* commented on this pull request.
>>
>> Just a number of minor stylistic comments now. Looks good otherwise.
>>
>> ------------------------------------------------------------------------
>>
>> In request.lisp 
>> <https://github.com/edicl/drakma/pull/75#pullrequestreview-6707176>:
>>
>> > -  "Helper function to read from stream into a buffer of element-type, which is returned."
>> -  (let ((buffer (make-array +buffer-size+ :element-type element-type))
>> -        (result (make-array 0 :element-type element-type :adjustable t)))
>> -        (loop for index = 0 then (+ index pos)
>> -           for pos = (read-sequence buffer stream)
>> -           do (adjust-array result (+ index pos))
>> -             (replace result buffer :start1 index :end2 pos)
>> -           while (= pos +buffer-size+))
>> -        result))
>> -
>> -(defun read-body (stream headers textp &key (decode-content t))
>> +  ;; In 32-bit Clozure-CL the maximum array size limit
>> +  ;; is (expt 2 24) elements.
>> +  ;; See:http://ccl.clozure.com/manual/chapter4.7.html
>> +  ;; Use a list as buffer
>> +  (progn
>>
>> This |progn| is no longer needed.
>>
>> ------------------------------------------------------------------------
>>
>> In request.lisp 
>> <https://github.com/edicl/drakma/pull/75#pullrequestreview-6707176>:
>>
>> > -           for pos = (read-sequence buffer stream)
>> -           do (adjust-array result (+ index pos))
>> -             (replace result buffer :start1 index :end2 pos)
>> -           while (= pos +buffer-size+))
>> -        result))
>> -
>> -(defun read-body (stream headers textp &key (decode-content t))
>> +  ;; In 32-bit Clozure-CL the maximum array size limit
>> +  ;; is (expt 2 24) elements.
>> +  ;; See:http://ccl.clozure.com/manual/chapter4.7.html
>> +  ;; Use a list as buffer
>> +  (progn
>> +    (when (>= buffer-size array-total-size-limit)
>> +      (setq buffer-size (- array-total-size-limit 1)))
>> +    (let ((buffer (make-array buffer-size :element-type element-type))
>> +          (result ()))
>>
>> Both |buffer| and |result| can be declared inside of the loop using 
>> |with|.
>>
>> ------------------------------------------------------------------------
>>
>> In request.lisp 
>> <https://github.com/edicl/drakma/pull/75#pullrequestreview-6707176>:
>>
>> > -           do (adjust-array result (+ index pos))
>> -             (replace result buffer :start1 index :end2 pos)
>> -           while (= pos +buffer-size+))
>> -        result))
>> -
>> -(defun read-body (stream headers textp &key (decode-content t))
>> +  ;; In 32-bit Clozure-CL the maximum array size limit
>> +  ;; is (expt 2 24) elements.
>> +  ;; See:http://ccl.clozure.com/manual/chapter4.7.html
>> +  ;; Use a list as buffer
>> +  (progn
>> +    (when (>= buffer-size array-total-size-limit)
>> +      (setq buffer-size (- array-total-size-limit 1)))
>> +    (let ((buffer (make-array buffer-size :element-type element-type))
>> +          (result ()))
>> +      (loop :for index = 0 :then (+ index pos)
>>
>> Please use bare symbols, not keywords, as loop keywords.
>>
>> ------------------------------------------------------------------------
>>
>> In request.lisp 
>> <https://github.com/edicl/drakma/pull/75#pullrequestreview-6707176>:
>>
>> > -        (result (make-array 0 :element-type element-type :adjustable t)))
>> -        (loop for index = 0 then (+ index pos)
>> -           for pos = (read-sequence buffer stream)
>> -           do (adjust-array result (+ index pos))
>> -             (replace result buffer :start1 index :end2 pos)
>> -           while (= pos +buffer-size+))
>> -        result))
>> -
>> -(defun read-body (stream headers textp &key (decode-content t))
>> +  ;; In 32-bit Clozure-CL the maximum array size limit
>> +  ;; is (expt 2 24) elements.
>> +  ;; See:http://ccl.clozure.com/manual/chapter4.7.html
>> +  ;; Use a list as buffer
>> +  (progn
>> +    (when (>= buffer-size array-total-size-limit)
>> +      (setq buffer-size (- array-total-size-limit 1)))
>>
>> Re-bind buffer size instead of using |setq|. Do so inside of the 
>> |loop| using |with| (see below for more of that).
>>
>> ------------------------------------------------------------------------
>>
>> In request.lisp 
>> <https://github.com/edicl/drakma/pull/75#pullrequestreview-6707176>:
>>
>> > -           while (= pos +buffer-size+))
>> -        result))
>> -
>> -(defun read-body (stream headers textp &key (decode-content t))
>> +  ;; In 32-bit Clozure-CL the maximum array size limit
>> +  ;; is (expt 2 24) elements.
>> +  ;; See:http://ccl.clozure.com/manual/chapter4.7.html
>> +  ;; Use a list as buffer
>> +  (progn
>> +    (when (>= buffer-size array-total-size-limit)
>> +      (setq buffer-size (- array-total-size-limit 1)))
>> +    (let ((buffer (make-array buffer-size :element-type element-type))
>> +          (result ()))
>> +      (loop :for index = 0 :then (+ index pos)
>> +            :for pos = (read-sequence buffer stream :start 0 :end buffer-size)
>> +            :do (dotimes (i pos) (push (aref buffer i) result))
>>
>> Put the |push| onto a separate line.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub 
>> <https://github.com/edicl/drakma/pull/75#pullrequestreview-6707176>, 
>> or mute the thread 
>> <https://github.com/notifications/unsubscribe-auth/APn5KN4Wo09h4wYtk-F3a7eB3wmU0gk2ks5q560xgaJpZM4KlMSF>.
>>
>

hanshuebner commented:
I don't quite see what you need the `setq` for.
```
(let ((result (loop ....)))
  (if (subtypep ...
```
Use three backticks around your code in comments to make it easier to review.

hanshuebner commented:
To use Markdown, you'll have to go through the GitHub web interface, it seems.

stassats commented:
You don't need a new unit test, the current tests already fail on returning a list.
You added a coercion to string, but now it fails by returning a list for :force-binary t.
This approaches the game of whack-a-mole. It should never return a list.

charcodelimit commented:
Could you provide me with some details on the test set-up you used?

My test results for SBCL 1.3.9 with  commit `7b071df0e4c0e3b4d7f2edbba04359b4fde40ee5`
look like this:

```
 Did 30 checks.
    Pass: 30 (100%)
    Skip: 0 ( 0%)
    Fail: 0 ( 0%)
;
; compilation unit finished
;   caught 1 STYLE-WARNING condition
#<ASDF/LISP-ACTION:TEST-OP >
#<ASDF/PLAN:SEQUENTIAL-PLAN {10080B5D53}>
```

stassats commented:
You don't need a test to know that %read-body should not return a list.
And I don't think this change is really worthwhile even if done right.

charcodelimit commented:
Just out of curiosity. 

Efficiency issues aside, what can you do in terms of the sequence protocol (http://www.lispworks.com/documentation/HyperSpec/Body/c_sequen.htm) with an array that you can't do with a list?
